### ISSUE TITLE
(Feature request) Adaptive step size for plots?

### ISSUE THREAD

**Issue by newbthenewbd** [Qalculate/libqalculate#420]
Bear with me if the title just so happens to make about as much sense as "electric chair for cold winters?", that's just my ability in math. :slightly_smiling_face:

So, **background to prevent XY:**
Let's say I want to plot the following simple ellipse: `1 = x^2/3 + y^2/4`

The `solve` function correctly and wowzeramazingly breaks it up into: `[(2 × √(1 − x^2 / 3))  (−2 × √(1 − x^2 / 3))]`

Now, I'd like to plot it... But this is the output of the ridiculously accurate `plot([(2 × sqrt(1 − ((x^2) / 3)))  ((−2) × sqrt(1 − ((x^2) / 3)))]; −10; 10; 0,0001; x; 1)` which takes... a while to compute:

![image](https://user-images.githubusercontent.com/1766353/160236155-53f8c1c9-8ac6-42cf-8023-174df4393b2b.png)

Zoomed to an area of interest:

![image](https://user-images.githubusercontent.com/1766353/160236242-b96430fc-2472-489f-a738-aba518e53660.png)

And it's only much, much worse with the default plotting settings:

![image](https://user-images.githubusercontent.com/1766353/160236456-f09b3051-396c-4968-aa3f-011a4871fca2.png)

It seems to me that one method to solve this problem (which, for all I know, doesn't occur in GeoGebra... not the correct measure, is it? (:) would be a step size that is somehow tied to the derivative of the function being plotted, resulting in these extreme steep places getting enough points.

Thanks for your attention!

**Comment by hanna-kn** [comment]
I agree adaptive step size is a good idea. I have just not found enough motivation to implement it yet.

You can avoid issues with the end points by specifying appropriate max/min x values. In this case this would ideally be -sqrt(3) and sqrt(3), but unfortunately plot points are approximated without trying to find an exact value (for efficiency reasons), which means that the y values for -sqrt(3) and sqrt(3) are complex and thus ignored. `plot([(2 × sqrt(1 − ((x^2) / 3))) ((−2) × sqrt(1 − ((x^2) / 3)))]; -1,732050807; 1,732050807)` does however result in a more complete circle.

**Comment by hanna-kn** [comment]
I implemented adaptive step size, using the derivative of the function as weight, but because the result in some case is worse, I've deactivated it for now. Instead I've improved the precision of real values before and after non-real values.

**Comment by idamir** [comment]
Please update pot files.

**Comment by hanna-kn** [comment]
I've now implemented a less aggressive adaptive spacing between x values (half of values are allocated).

**Comment by newbthenewbd** [comment]
Apologies for letting the issue just hang there, confirming that this is now royally awesome even with no plot precision tweaking at all :) Contrary to what my «responsiveness» might suggest, I've actually been heavily using Qalculate! this whole time, it's state-of-the-art and yet improving! Thank you so much!

![image](https://github.com/Qalculate/libqalculate/assets/1766353/0af9bcb9-f632-4f81-8f76-58a3a4e12578)
